### PR TITLE
feat: Provide support for KustoClusters and NatGateways resources

### PR DESCRIPTION
**Pull request by SagiVaknin**
Relates to https://github.com/tomkerkhove/promitor/issues/2225
Relates to https://github.com/tomkerkhove/promitor/issues/340

### CODE REVIEW

File: src/Promitor.Agents.ResourceDiscovery/Graph/ResourceTypes/KustoClusterDiscoveryQuery.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using GuardNet;
+using Newtonsoft.Json.Linq;
+using Promitor.Core.Contracts;
+using Promitor.Core.Contracts.ResourceTypes;
+
+namespace Promitor.Agents.ResourceDiscovery.Graph.ResourceTypes
+{
+    public class KustoClusterDiscoveryQuery : ResourceDiscoveryQuery
```
Comment (tomkerkhove) [inline src/Promitor.Agents.ResourceDiscovery/Graph/ResourceTypes/KustoClusterDiscoveryQuery.cs:8]: This is more of a general remark, but we should rename all `Kusto` references to `DataExplorer` given that is the product name.
```suggestion
    public class DataExplorerClusterDiscoveryQuery : ResourceDiscoveryQuery
```

File: config/promitor/scraper/metrics.yaml [hunk 2]
```diff
@@ -312,6 +312,24 @@ metrics:
...
+    resourceType: NatGateway
+    azureMetricConfiguration:
+      metricName: DatapathAvailability
+      aggregation:
+        type: Average
+    resourceDiscoveryGroups:
+      - name: nat-gateways
+  - name: promitor_demo_mysql_kusto_cluster_CPU
+    description: "CPU usage of kusto cluster"
+    resourceType: KustoCluster
```
Comment (tomkerkhove) [inline config/promitor/scraper/metrics.yaml:326]:
```suggestion
    resourceType: DataExplorerCluster
```

File: src/Promitor.Agents.Scraper/Validation/MetricDefinitions/ResourceTypes/KustoClusterMetricValidator.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public IEnumerable<string> Validate(MetricDefinition metricDefinition)
+        {
+            Guard.NotNull(metricDefinition, nameof(metricDefinition));
+
+            foreach (var resourceDefinition in metricDefinition.Resources.Cast<KustoClusterResourceDefinition>())
+            {
+                if (string.IsNullOrWhiteSpace(resourceDefinition.KustoClusterName))
+                {
+                    yield return "No kusto cluster name is configured";
```
Comment (tomkerkhove) [inline src/Promitor.Agents.Scraper/Validation/MetricDefinitions/ResourceTypes/KustoClusterMetricValidator.cs:20]:
```suggestion
                    yield return "No Azure Data Explorer cluster name is configured";
```

File: src/Promitor.Agents.Scraper/Validation/MetricDefinitions/ResourceTypes/NatGatewayMetricValidator.cs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public IEnumerable<string> Validate(MetricDefinition metricDefinition)
+        {
+            Guard.NotNull(metricDefinition, nameof(metricDefinition));
+
+            foreach (var resourceDefinition in metricDefinition.Resources.Cast<NatGatewayResourceDefinition>())
+            {
+                if (string.IsNullOrWhiteSpace(resourceDefinition.NatGatewayName))
+                {
+                    yield return "No Azure NAT Gateway name is configured";
```
Comment (tomkerkhove) [inline src/Promitor.Agents.Scraper/Validation/MetricDefinitions/ResourceTypes/NatGatewayMetricValidator.cs:20]: Nit - Let's align with product naming
```suggestion
                    yield return "No Azure Virtual Network NAT Gateway name is configured";
```

File: src/Promitor.Core.Contracts/ResourceType.cs [hunk 5]
```diff
@@ -50,5 +50,7 @@ public enum ResourceType
         LoadBalancer = 45,
         MySql = 46,
         LogAnalytics = 47,
+        KustoCluster = 48,
```
Comment (tomkerkhove) [inline src/Promitor.Core.Contracts/ResourceType.cs:53]:
```suggestion
        DataExplorerCluster = 48,
```

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Model/ResourceTypes/KustoClusterResourceV1.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+{
+    /// <summary>
+    /// Contains the configuration required to scrape a kusto cluster.
+    /// </summary>
+    public class KustoClusterResourceV1 : AzureResourceDefinitionV1
+    {
+        /// <summary>
+        /// The name of the Kusto Cluster to get metrics for.
+        /// </summary>
+        public string KustoClusterName { get; set; }
```
Comment (tomkerkhove) [inline src/Promitor.Core.Scraping/Configuration/Serialization/v1/Model/ResourceTypes/KustoClusterResourceV1.cs:11]:
```suggestion
        public string ClusterName { get; set; }
```

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Model/ResourceTypes/KustoClusterResourceV1.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Promitor.Core.Scraping.Configuration.Serialization.v1.Model.ResourceTypes
+{
+    /// <summary>
+    /// Contains the configuration required to scrape a kusto cluster.
+    /// </summary>
+    public class KustoClusterResourceV1 : AzureResourceDefinitionV1
```
Comment (tomkerkhove) [inline src/Promitor.Core.Scraping/Configuration/Serialization/v1/Model/ResourceTypes/KustoClusterResourceV1.cs:6]:
```suggestion
    public class DataExplorerClusterResourceV1 : AzureResourceDefinitionV1
```

File: src/Promitor.Tests.Unit/Serialization/v1/Providers/KustoClusterDeserializerTests.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            YamlAssert.PropertySet<KustoClusterResourceV1, AzureResourceDefinitionV1, string>(
+                _deserializer,
+                "kustoClusterName: promitor-kusto-cluster",
+                "promitor-kusto-cluster",
+                r => r.KustoClusterName);
+        }
+
+        [Fact]
+        public void Deserialize_KustoClusterNameNotSupplied_Null()
```
Comment (tomkerkhove) [inline src/Promitor.Tests.Unit/Serialization/v1/Providers/KustoClusterDeserializerTests.cs:31]: Can we add a test similar to Deserialize_NatGatewayNameNotSupplied_ReportsError please?

**Review by tomkerkhove — CHANGES_REQUESTED** [review]
LGTM, but a few small remarks:
- Can you please update the changelog? ([link](https://github.com/tomkerkhove/promitor/blob/master/CONTRIBUTING.md#changelog))
- We should rename `KustoCluster` to `DataExplorerCluster` as that is the general product name and resonates more with end-users

**Comment by tomkerkhove** [comment]
And thank you very much for the PR btw!

File: src/Promitor.Core.Contracts/ResourceType.cs (hunk 5 above)
Comment (SagiVaknin) [reply src/Promitor.Core.Contracts/ResourceType.cs]: Renamed to data explorer

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Model/ResourceTypes/KustoClusterResourceV1.cs (hunk 6 above)
Comment (SagiVaknin) [reply src/Promitor.Core.Scraping/Configuration/Serialization/v1/Model/ResourceTypes/KustoClusterResourceV1.cs]: Renamed

File: src/Promitor.Core.Scraping/Configuration/Serialization/v1/Model/ResourceTypes/KustoClusterResourceV1.cs (hunk 7 above)
Comment (SagiVaknin) [reply src/Promitor.Core.Scraping/Configuration/Serialization/v1/Model/ResourceTypes/KustoClusterResourceV1.cs]: Renamed

File: src/Promitor.Agents.Scraper/Validation/MetricDefinitions/ResourceTypes/NatGatewayMetricValidator.cs (hunk 4 above)
Comment (SagiVaknin) [reply src/Promitor.Agents.Scraper/Validation/MetricDefinitions/ResourceTypes/NatGatewayMetricValidator.cs]: Added proper product name

File: src/Promitor.Tests.Unit/Serialization/v1/Providers/KustoClusterDeserializerTests.cs (hunk 8 above)
Comment (SagiVaknin) [reply src/Promitor.Tests.Unit/Serialization/v1/Providers/KustoClusterDeserializerTests.cs]: Added a similar test to DataExplorer

File: src/Promitor.Agents.ResourceDiscovery/Graph/ResourceTypes/KustoClusterDiscoveryQuery.cs (hunk 1 above)
Comment (SagiVaknin) [reply src/Promitor.Agents.ResourceDiscovery/Graph/ResourceTypes/KustoClusterDiscoveryQuery.cs]: Renamed respectively.

File: config/promitor/scraper/metrics.yaml (hunk 2 above)
Comment (SagiVaknin) [reply config/promitor/scraper/metrics.yaml]: Renamed

**Comment by SagiVaknin** [comment]
> And thank you very much for the PR btw!

👍  Thanks for the project :)

File: src/Promitor.Agents.ResourceDiscovery/Graph/ResourceDiscoveryFactory.cs [hunk 9]
```diff
@@ -30,6 +30,8 @@ public static ResourceDiscoveryQuery UseResourceDiscoveryFor(ResourceType resour
                     return new CosmosDbDiscoveryQuery();
                 case ResourceType.DataFactory:
                     return new DataFactoryDiscoveryQuery();
+                case ResourceType.DataExplorerCluster:
```
Comment (tomkerkhove) [inline src/Promitor.Agents.ResourceDiscovery/Graph/ResourceDiscoveryFactory.cs:33]: Nit - Can you alphabetically sort please?

File: src/Promitor.Agents.Scraper/Validation/Factories/MetricValidatorFactory.cs [hunk 10]
```diff
@@ -33,6 +33,8 @@ internal static IMetricValidator GetValidatorFor(ResourceType resourceType)
                     return new CosmosDbMetricValidator();
                 case ResourceType.DataFactory:
                     return new DataFactoryMetricValidator();
+                case ResourceType.DataExplorerCluster:
```
Comment (tomkerkhove) [inline src/Promitor.Agents.Scraper/Validation/Factories/MetricValidatorFactory.cs:36]: Nit - Can you alphabetically sort please?

**Review by tomkerkhove — COMMENTED** [review]
Nice, thank you! Just added two nits.
Can you also provide the other 2 things from https://github.com/tomkerkhove/promitor/pull/2224#pullrequestreview-1295204879 please?

**Comment by tomkerkhove** [comment]
Any update on this @SagiVaknin?

File: src/Promitor.Agents.ResourceDiscovery/Graph/ResourceDiscoveryFactory.cs (hunk 9 above)
Comment (SagiVaknin) [reply src/Promitor.Agents.ResourceDiscovery/Graph/ResourceDiscoveryFactory.cs:31]: Fixed nit

File: src/Promitor.Agents.Scraper/Validation/Factories/MetricValidatorFactory.cs (hunk 10 above)
Comment (SagiVaknin) [reply src/Promitor.Agents.Scraper/Validation/Factories/MetricValidatorFactory.cs:34]: Fixed nit

**Comment by SagiVaknin** [comment]
> Any update on this @SagiVaknin?

Didn't had time to look at it, will also update the docs

File: changelog/content/experimental/unreleased.md [hunk 11]
```diff
@@ -10,10 +10,12 @@ version:
 - {{% tag added %}} Provide capability to read Azure AD service principal secret from a file
 - {{% tag added %}} Provide Azure Log Analytics scraper ([docs](https://docs.promitor.io/v2.9/scraping/providers/log-analytics/)
 | [#2132](https://github.com/tomkerkhove/promitor/pull/2132))
+- {{% tag added %}} Provide support for Azure NAT Gateway and Azure Data Explorer Clusters.
```
Comment (tomkerkhove) [inline changelog/content/experimental/unreleased.md:13]: THis should be separate lines
```suggestion
- {{% tag added %}} Provide support for Azure Data Explorer Clusters. ([docs](https://docs.promitor.io/v2.9/scraping/providers/data-explorer-clusters.md)
- {{% tag added %}} Provide support for Azure NAT Gateway. ([docs](https://docs.promitor.io/v2.9/scraping/providers/nat-gateway.md)
```

File: changelog/content/experimental/unreleased.md [hunk 12]
```diff
@@ -10,10 +10,12 @@ version:
...
 - {{% tag fixed %}} Fixed a bug where startup throws scheduling exception due to metric misconfiguration
 - {{% tag fixed %}} Fixed a bug where resource discovery for Azure Container Instances was not working
 - {{% tag fixed %}} Fixed a bug where Promitor was reported as `unknown_service:dotnet` job in OpenTelemetry Collector
 
 #### Resource Discovery
 
 - {{% tag added %}} Provide path to read app secret key from file
+- {{% tag added %}} Provide support for Azure NAT Gateway and Azure Data Explorer Clusters.
```
Comment (tomkerkhove) [inline changelog/content/experimental/unreleased.md:21]:
```suggestion
- {{% tag added %}} Provide support for Azure Data Explorer Clusters. ([docs](https://docs.promitor.io/v2.9/scraping/providers/data-explorer-clusters.md)
- {{% tag added %}} Provide support for Azure NAT Gateway. ([docs](https://docs.promitor.io/v2.9/scraping/providers/nat-gateway.md)
```

**Review by tomkerkhove — APPROVED** [review]

**Comment by tomkerkhove** [comment]
@SagiVaknin It looks like you have another users that has to sign the CLA; would you mind going ahead and doing that please?

**Comment by tomkerkhove** [comment]
Closing in favor of #2250

**Pull request closed without merging**